generators/copy.go

```diff
@@ -0,0 +1,218 @@
+func (g CopyGenerator) Run(cacheDir, sourceDir string,
+	defFile shared.DefinitionFile) error {
+
+	// First check if the input is a file or a directory.
+	// Then check whether the destination finishes in a "/" or not
+	// Afterwards, the rules for copying can be applied
+	// Rules for copying:
+	// - If <src> is a directory, the entire contents of the directory
+	// are copied. Only symlinks and regular files are supported.
+	// Note 1: The directory itself is not copied, just its contents.
+	// Note 2: For files copied, only regular unix permissions are kept
+	// - If <src> is any other kind of file, it is copied individually
+	// along with its metadata. In this case, if <dest> ends with a trailing
+	// slash /, it will be considered a directory and the contents of <src>
+	// will be written at <dest>/base(<src>).
+	// - If <dest> does not end with a trailing slash, it will be considered
+	// a regular file and the contents of <src> will be written at <dest>.
+	// - If <dest> doesn’t exist, it is created along with all missing
+	// directories in its path.
+	// - Multiple <src> resources can be specified using golang
+	// filepath.Match regexps.
+	// For simplicity they are only allowed in the basename and not in
+	// the directory hierarchy. If more than one match is found, <dest>
+	// will be automatically interpreted as a directory
+
+	// First let's set the name of the destination file to the input file
+	// relative to the root if it is missing
+	var destPath, srcPath string
+	var files []string
+	srcPath = defFile.Source
+	destPath = filepath.Join(sourceDir, defFile.Source)
+	if defFile.Path != "" {
+		destPath = filepath.Join(sourceDir, defFile.Path)
+	}
+
+	dirFiles, err := ioutil.ReadDir(filepath.Dir(srcPath))
+	if err != nil {
+		return err
+	}
+	for _, f := range dirFiles {
+		match, err := filepath.Match(srcPath, filepath.Join(filepath.Dir(srcPath), f.Name()))
+		if err != nil {
+			return err
+		}
+		if match {
+			files = append(files, filepath.Join(filepath.Dir(srcPath), f.Name()))
+		}
+	}
+
+	switch len(files) {
+	case 0:
+		// Look for the literal file
+		_, err = os.Stat(srcPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				err = fmt.Errorf("File '%s' doesn't exist", srcPath)
+			}
+			return err
+		}
+		err = copy(srcPath, destPath, defFile)
+	case 1:
+		err = copy(srcPath, destPath, defFile)
+	default:
+		// Make sure that we are copying to a directory
+		defFile.Path = defFile.Path + "/"
+		for _, f := range files {
+			err = copy(f, destPath, defFile)
+			if err != nil {
+				break
+			}
+		}
+	}
+	return err
+}
+
+func copy(srcPath, destPath string, defFile shared.DefinitionFile) error {
+	in, err := os.Stat(srcPath)
+	if err != nil {
+		return err
+	}
+
+	switch in.Mode() & os.ModeType {
+	// Regular file
+	case 0:
+		if strings.HasSuffix(defFile.Path, "/") {
+			destPath = filepath.Join(destPath, filepath.Base(srcPath))
+		} else {
+		}
+		err := copyFile(srcPath, destPath, defFile)
```

Comment:
Could use `lxd.FileCopy` or `lxd.DirCopy` there I think.

(`lxd` being the name we use for imports of `github.com/lxc/lxd/shared` inside the distrobuilder codebase)

Reply:
lxd.FileCopy would fit nicely in there, but also everywhere else where copyFile is used. In addition, it has support for symlinks, so that would increase functionality. However, I believe that keeping copyFile  as a wrapper for lxd.FileCopy makes sense to:
 * Make sure that we create the destination directory if missing
 * Run updateFileAccess to keep uid, gid and mode options functional

Does that make sense to you?

Reply:
Yeah, that should be fine

Reply:
Ok, it's added now and also a symlink has been added to the tests.